src/utils/recipeLoader.ts

```diff
@@ -173,4 +174,135 @@ const resolveRefs = (
     return viewable;
 }
 
-export { isFirebaseRef, resolveRefs, isInRefsByCollection, addRef };
\ No newline at end of file
+const getRecipeDoc = async (id: string): Promise<FirebaseRecipe> => {
+    const querySnapshot = await queryDocumentById(FIRESTORE_COLLECTIONS.RECIPES, id);
+    const docs: Array<FirebaseRecipe> = querySnapshot.docs.map((doc) => ({
+        id: doc.id,
+        name: doc.data().name,
+        dedup_hash: doc.data().dedup_hash,
+        ...doc.data(),
+    }));
+    if (docs.length === 0) {
+        throw new Error(`Recipe with ID ${id} not found`);
+    }
+    return docs[0];
+};
+
+const searchForRefs = async (
+    refs: string[],
+    refsToObj: RefsByCollection
+): Promise<RefsByCollection> => {
+    // split up the refs we're searching for by collection, so we can
+    // handle searching with one query per collection
+    const refsToGetByCollection: Dictionary<Array<string>> = {};
+    refs.forEach((ref) => {
+        const splitRef: string[] = ref.split((/:|\//));
+        const collectionName = splitRef[1];
+        // Only need to search if we don't already have the object for this reference
+        if (!isInRefsByCollection(ref, collectionName, refsToObj)) {
+            // firebase:collection/doc -> [firebase, collection, doc]
+            if (!(collectionName in refsToGetByCollection)) {
+                refsToGetByCollection[collectionName] = [];
+            }
+            if (!(splitRef[2] in refsToGetByCollection[collectionName])) {
+                // Only need to search for each ref once
+                refsToGetByCollection[collectionName].push(splitRef[2]);
```

Comment:
Oh i somehow just realized that this `in` checks indices which would cause a duplicate push. We might want to switch to using membership check: `if (!refsToGetByCollection[collectionName].includes(splitRef[2]))`